### ISSUE TITLE
BUG: closure is created for (this) when nameof() is referencing a member

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#53591]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/BUG:-closure-is-created-for-this-when-/1427334)._

---
Visual Studio 16.9.6

When an anonymous method contains a reference to a member in a nameof() expression, the "this" variable gets captured. This can be circumvented by fully specifying the member, however i think the capture should never be created.

```
void ClosureCreated()
{
  Func<string> x = () => nameof(ClosureCreated);
}
```

![image](https://user-images.githubusercontent.com/1408396/119170487-248af500-ba18-11eb-888d-3e21dbab951f.png)

---
#### Feedback Bot on 5/19/2021, 10:45 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

**Comment by RikkiGibson** [comment]
worth noting that this is just a dataflow API issue. Putting `static` on the lambda for example compiles without errors and the emitted code doesn't take a reference to `class Example`.

**Comment by jinujoseph** [comment]
cc @jcouv

**Comment by jcouv** [comment]
Moving the issue to Area-compilers since is a problem with the compiler data flow API.

**Comment by jcouv** [comment]
@RikkiGibson Please let me know if you agree with my assessment below.

The definite assignment analysis still counts usages inside `nameof` (marked as unreachable in `AbstractFlowPass.VisitNameOfOperator`) and other unreachable expressions. I believe this is by design (it's meant to be conservative).
If we didn't do that (see experiment PR https://github.com/dotnet/roslyn/pull/55724) we would break useful scenarios:
```
        public void NameofLocal()
        {
            var source = @"
int local1;
int local2 = 0;
_ = nameof(local1);
_ = nameof(local2);
";
            CreateCompilation(source).VerifyDiagnostics(
                // (2,5): warning CS0168: The variable 'local1' is declared but never used
                // int local1;
                Diagnostic(ErrorCode.WRN_UnreferencedVar, "local1").WithArguments("local1").WithLocation(2, 5),
                // (3,5): warning CS0219: The variable 'local2' is assigned but its value is never used
                // int local2 = 0;
                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "local2").WithArguments("local2").WithLocation(3, 5)
                );
        }
```

**Comment by jcouv** [comment]
Brainstormed a possible design with Rikki. Will try that. Thanks for the suggestion!

**Comment by jcouv** [comment]
Fix merged for 17.0p4
